lib/optimizely/custom_attribute_condition_evaluator.rb

```diff
@@ -0,0 +1,171 @@
+# frozen_string_literal: true
+
+#
+#    Copyright 2018, Optimizely and contributors
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+module Optimizely
+  class CustomAttributeConditionEvaluator
+    CUSTOM_ATTRIBUTE_CONDITION_TYPE = 'custom_attribute'
+
+    # Conditional match types
+    EXACT_MATCH_TYPE = 'exact'
+    EXISTS_MATCH_TYPE = 'exists'
+    GREATER_THAN_MATCH_TYPE = 'gt'
+    LESS_THAN_MATCH_TYPE = 'lt'
+    SUBSTRING_MATCH_TYPE = 'substring'
+
+    EVALUATORS_BY_MATCH_TYPE = {
+      EXACT_MATCH_TYPE => :exact_evaluator,
+      EXISTS_MATCH_TYPE => :exists_evaluator,
+      GREATER_THAN_MATCH_TYPE => :greater_than_evaluator,
+      LESS_THAN_MATCH_TYPE => :less_than_evaluator,
+      SUBSTRING_MATCH_TYPE => :substring_evaluator
+    }.freeze
+
+    FINITE_NUMBER_LIMIT = 1.0e+53
+
+    attr_reader :user_attributes
+
+    def initialize(user_attributes)
+      @user_attributes = user_attributes
+    end
+
+    def evaluate(leaf_condition)
+      # Top level method to evaluate audience conditions.
+      #
+      # conditions - Nested array of and/or conditions.
+      #              Example: ['and', operand_1, ['or', operand_2, operand_3]]
+      #
+      # Returns boolean if the given user attributes match/don't match the given conditions,
+      #         nil if the given conditions can't be evaluated.
+
+      return nil unless leaf_condition['type'] == CUSTOM_ATTRIBUTE_CONDITION_TYPE
+
+      condition_match = leaf_condition['match']
+
+      return nil if !condition_match.nil? && !EVALUATORS_BY_MATCH_TYPE.include?(condition_match)
```

Comment:
You can avoid checking !condition_match.nil? is you keep the below `condition_match = EXACT_MATCH_TYPE if condition_match.nil?` before this

Reply:
This is to handle invalid match types cases.